building_sim_plugins/building_gazebo_plugins/src/crowd_simulator.cpp

```diff
@@ -0,0 +1,499 @@
+    std::vector<std::string> parts(
+      std::sregex_token_iterator(poseStr.begin(), poseStr.end(), ws_re, -1),
+      std::sregex_token_iterator());
+
+    if (parts.size() != 6)
+    {
+      gzerr <<
+        "Error loading <initial_pose> in <model_type>, 6 floats (x, y, z, pitch, roll, yaw) expected.";
+      return false;
+    }
+
+    double x = ignition::math::parseFloat(parts[0]);
+    double y = ignition::math::parseFloat(parts[1]);
+    double z = ignition::math::parseFloat(parts[2]);
+    double pitch = ignition::math::parseFloat(parts[3]);
+    double roll = ignition::math::parseFloat(parts[4]);
+    double yaw = ignition::math::parseFloat(parts[5]);
+
+    result.X() = x;
+    result.Y() = y;
+    result.Z() = z;
+    result.Pitch() = pitch;
+    result.Roll() = roll;
+    result.Yaw() = yaw;
+  }
+  return true;
+}
+
+
+//============================================
+bool CrowdSimulatorPlugin::_CreateModel(const std::string& modelName,
+  const crowd_simulator::ModelTypeDatabase::Record* modelTypePtr,
+  const crowd_simulator::AgentPtr agentPtr)
+{
+  sdf::ElementPtr modelElement(new sdf::Element());
+  modelElement->SetName("actor");
+  modelElement->AddAttribute("name", "string", modelName, true);
+
+  sdf::ElementPtr skinElement(new sdf::Element());
+  skinElement->SetName("skin");
+  modelElement->InsertElement(skinElement);
+
+  sdf::ElementPtr animElement(new sdf::Element());
+  animElement->SetName("animation");
+  animElement->AddAttribute("name", "string", "walk", true);
+  modelElement->InsertElement(animElement);
+
+  sdf::ElementPtr filenameElement(new sdf::Element());
+  filenameElement->SetName("filename");
+  filenameElement->AddValue("string", modelTypePtr->fileName, true);
+  skinElement->InsertElement(filenameElement);
+  animElement->InsertElement(filenameElement);
+
+  sdf::ElementPtr poseElement(new sdf::Element());
+  poseElement->SetName("pose");
+  std::ostringstream oss;
+
+  oss << agentPtr->_pos.x() << " " << agentPtr->_pos.y() << " " << "0 0 0 0";
+  poseElement->AddValue("pose", oss.str(), true);
+  modelElement->InsertElement(poseElement);
+
+  sdf::SDFPtr sdf(new sdf::SDF());
+  sdf->Root(modelElement);
+
+  assert(sdf);
+  this->_world->InsertModelSDF(*sdf);
+  gzdbg << "Insert corresponding model for simulator agent: [" << modelName <<
+    " ]." << std::endl;
+  return true;
+}
+
+//===================================================
+
+crowd_simulator::AgentPose3d Convert(const ignition::math::Pose3d& ignition_pose){
+  auto pos = ignition_pose.Pos();
+  auto euler = ignition_pose.Rot().Euler();
+
+  return crowd_simulator::AgentPose3d(pos.X(), pos.Y(), pos.Z(), euler.X(), euler.Y(), euler.Z());
+}
+
+ignition::math::Pose3d Convert(const crowd_simulator::AgentPose3d& agent_pose){
+  return ignition::math::Pose3d(
+    agent_pose.X(), 
+    agent_pose.Y(), 
+    agent_pose.Z(), 
+    agent_pose.Pitch(), 
+    agent_pose.Roll(), 
+    agent_pose.Yaw());
+}
```

Comment:
Could these functions be moved to the common library?

Reply:
The purpose of putting the conversion in here is I want the common library independent with ignition lib. Because the gazebo plugin and the ignition plugin are using different version of ign lib. Like other plugins are using eigen as the connection. But other plugins are one-way connection. crowd_sim plugin has two-way for external_agent and other agent. So I think if the gazebo plugin and ign plugin are continuing use different ign lib versions, this function might need to keep, separately.... though I don't know for sure.

Reply:
I think you can do a templated function, similar to what is done in [slotcar_common](https://github.com/FloodShao/traffic_editor/blob/master/building_sim_plugins/building_plugins_common/include/building_sim_common/slotcar_common.hpp#L19-L26), then the header wouldn't have an ignition dependency

Reply:
thanks for the notice, will modify it.